src/ocaml-output/FStar_SMTEncoding_SplitQueryCases.ml

```diff
@@ -1,132 +1,180 @@
+          FStar_SMTEncoding_Term.term))
+  =
+  fun n  ->
+    fun t  ->
+      fun f  ->
+        match t.FStar_SMTEncoding_Term.tm with
+        | FStar_SMTEncoding_Term.Quant
+            (FStar_SMTEncoding_Term.Forall ,l,opt,l',t) ->
+            parse_query_for_split_cases n t
+              (fun x  ->
+                 f (FStar_SMTEncoding_Util.mkForall'' (l, opt, l', x)))
+        | FStar_SMTEncoding_Term.App
+            (FStar_SMTEncoding_Term.Imp ,t1::t2::uu____163) ->
+            let r =
+              match t2.FStar_SMTEncoding_Term.tm with
+              | FStar_SMTEncoding_Term.Quant
+                  (FStar_SMTEncoding_Term.Forall
+                   ,uu____183,uu____184,uu____185,uu____186)
+                  ->
+                  parse_query_for_split_cases n t2
+                    (fun x  -> f (FStar_SMTEncoding_Util.mkImp (t1, x)))
+              | FStar_SMTEncoding_Term.App
+                  (FStar_SMTEncoding_Term.ITE ,uu____196) ->
+                  let uu____199 =
+                    is_ite_all_the_way n t2 FStar_SMTEncoding_Util.mkTrue [] in
+                  (match uu____199 with
+                   | (b,l,negs) ->
+                       (b,
+                         (((fun x  ->
+                              f (FStar_SMTEncoding_Util.mkImp (t1, x)))), l,
+                           negs)))
+              | uu____227 ->
+                  (false,
+                    (((fun uu____237  -> FStar_SMTEncoding_Util.mkFalse)),
+                      [], FStar_SMTEncoding_Util.mkFalse)) in
+            r
+        | FStar_SMTEncoding_Term.App (FStar_SMTEncoding_Term.ITE ,uu____238)
+            ->
+            let uu____241 =
+              is_ite_all_the_way n t FStar_SMTEncoding_Util.mkTrue [] in
+            (match uu____241 with | (b,l,negs) -> (b, (f, l, negs)))
+        | uu____268 ->
+            (false,
+              (((fun uu____278  -> FStar_SMTEncoding_Util.mkFalse)), [],
+                FStar_SMTEncoding_Util.mkFalse))
+let strip_not: FStar_SMTEncoding_Term.term -> FStar_SMTEncoding_Term.term =
+  fun t  ->
+    match t.FStar_SMTEncoding_Term.tm with
+    | FStar_SMTEncoding_Term.App (FStar_SMTEncoding_Term.Not ,hd::uu____283)
+        -> hd
+    | uu____286 -> t
+let rec check_split_cases:
+  (FStar_SMTEncoding_Term.term -> FStar_SMTEncoding_Term.term) ->
+    FStar_SMTEncoding_Term.term Prims.list ->
+      (FStar_SMTEncoding_Term.decl -> Prims.unit) -> Prims.unit
+  =
+  fun f  ->
+    fun l  ->
+      fun check  ->
+        FStar_List.iter
+          (fun t  ->
+             check
+               (FStar_SMTEncoding_Term.Assume
+                  (let _0_358 = FStar_SMTEncoding_Util.mkNot (f t) in
+                   (_0_358, None, None)))) (FStar_List.rev l)
+let check_exhaustiveness:
+  (FStar_SMTEncoding_Term.term -> FStar_SMTEncoding_Term.term) ->
+    FStar_SMTEncoding_Term.term ->
+      (FStar_SMTEncoding_Term.decl -> Prims.unit) -> Prims.unit
+  =
+  fun f  ->
```

Comment:
I assume you've bootstrapped with your new printer as a showcase. Can we flatten these into n-ary funs?